File: rust/src/wiki.rs
use std::collections::BTreeMap;

use crate::domain::{RepoScan, WikiPage, WikiSection, WikiStructure};

pub fn plan_wiki(scan: &RepoScan, comprehensive: bool) -> WikiStructure {
    let title = repository_name(scan)
        .map(|name| format!("{name} Wiki"))
        .unwrap_or_else(|| "Project Wiki".to_string());
    let description = build_description(scan);

    let mut sections = Vec::new();
    let mut root_sections = Vec::new();
    let mut pages = Vec::new();

    let categorized = categorize_files(&scan.file_tree);
    let mut page_counter = 1usize;
    let mut section_counter = 1usize;

    for (category, file_paths) in categorized {
        if file_paths.is_empty() {
            continue;
        }

        let section_id = format!("section-{section_counter}");
        section_counter += 1;
        let page_id = format!("page-{page_counter}");
        page_counter += 1;

        let title_case = category_title(&category);
        pages.push(WikiPage {
            id: page_id.clone(),
            title: title_case.clone(),
            description: format!("Overview of the project's {} implementation.", title_case.to_lowercase()),
            importance: importance_for_category(&category).to_string(),
            file_paths: trim_files(file_paths, comprehensive),
            related_pages: Vec::new(),
            parent_section: Some(section_id.clone()),
            content: None,
        });

        sections.push(WikiSection {
            id: section_id.clone(),
            title: title_case,
            pages: vec![page_id],
            subsections: Vec::new(),
        });
        root_sections.push(section_id);
    }

    if pages.is_empty() {
        pages.push(WikiPage {
            id: "page-1".to_string(),
            title: "Repository Overview".to_string(),
            description: "High-level documentation page for the repository.".to_string(),
            importance: "high".to_string(),
            file_paths: trim_files(scan.file_tree.clone(), comprehensive),
            related_pages: Vec::new(),
            parent_section: None,
            content: None,
        });
    }

    let related_page_ids: Vec<String> = pages.iter().map(|page| page.id.clone()).collect();
    for page in &mut pages {
        page.related_pages = related_page_ids
            .iter()
            .filter(|id| **id != page.id)
            .take(3)
            .cloned()
            .collect();
    }

    WikiStructure {
        id: "wiki".to_string(),
        title,
        description,
        sections,
        root_sections,
        pages,
    }
}

fn categorize_files(files: &[String]) -> BTreeMap<String, Vec<String>> {
    let mut categories: BTreeMap<String, Vec<String>> = BTreeMap::new();

    for file in files {
        let category = if file.starts_with("tests/") || file.starts_with("test/") {
            "testing"
        } else if file.starts_with("src/app/") || file.starts_with("src/components/") || file.ends_with(".tsx") || file.ends_with(".jsx") {
            "frontend"
        } else if file.starts_with("api/") || file.ends_with(".py") {
            "backend"
        } else if file.contains("config") || file.ends_with(".json") || file.ends_with(".toml") || file.ends_with(".yaml") || file.ends_with(".yml") {
            "configuration"
        } else if file.contains("docker") || file.contains("compose") || file.contains("deploy") {
            "deployment"
        } else if file.ends_with(".md") {
            "documentation"
        } else {
            "architecture"
        };

        categories
            .entry(category.to_string())
            .or_default()
            .push(file.clone());
    }

    categories
}

fn trim_files(files: Vec<String>, comprehensive: bool) -> Vec<String> {
    let limit = if comprehensive { 12 } else { 6 };
    files.into_iter().take(limit).collect()
}

fn category_title(category: &str) -> String {
    match category {
        "frontend" => "Frontend Components".to_string(),
        "backend" => "Backend Systems".to_string(),
        "testing" => "Testing Strategy".to_string(),
        "configuration" => "Configuration".to_string(),
        "deployment" => "Deployment and Infrastructure".to_string(),
        "documentation" => "Project Documentation".to_string(),
        _ => "System Architecture".to_string(),
    }
}

fn importance_for_category(category: &str) -> &'static str {
    match category {
        "architecture" | "backend" | "frontend" => "high",
        "configuration" | "deployment" => "medium",
        _ => "low",
    }
}

fn repository_name(scan: &RepoScan) -> Option<String> {
    scan.root_path
        .file_name()
        .map(|name| name.to_string_lossy().to_string())
}

fn build_description(scan: &RepoScan) -> String {
    if let Some(readme) = &scan.readme_content {
        let summary = readme
            .lines()
            .map(str::trim)
            .find(|line| {
                !line.is_empty()
                    && !line.starts_with('#')
                    && !line.starts_with("---")
                    && !line.starts_with("![")
                    && !line.starts_with("[![")
            })
            .unwrap_or("Repository documentation generated from source structure.");
        return summary.to_string();
    }

    "Repository documentation generated from source structure.".to_string()
}

pub fn render_page_markdown(page: &WikiPage, scan: &RepoScan) -> String {
    let mut markdown = String::new();
    markdown.push_str("<details>\n<summary>Relevant source files</summary>\n\n");
    for file in &page.file_paths {
        markdown.push_str(&format!("- `{file}`\n"));
    }
    markdown.push_str("</details>\n\n");
    markdown.push_str(&format!("# {}\n\n", page.title));
    markdown.push_str(&format!("{}\n\n", page.description));

    markdown.push_str("## Summary\n\n");
    markdown.push_str(&format!(
        "This page was generated from the repository scan for `{}`. It currently serves as a structured Rust-owned wiki stub and lists the source files that should drive future LLM-backed generation.\n\n",
        scan.root_path.display()
    ));

    if let Some(branch) = &scan.default_branch {
        markdown.push_str("## Repository Metadata\n\n");
        markdown.push_str(&format!("- Default branch: `{branch}`\n"));
        markdown.push_str(&format!("- Scanned files: `{}`\n\n", scan.file_tree.len()));
    }

    markdown.push_str("## Source Inventory\n\n");
    for file in &page.file_paths {
        markdown.push_str(&format!("- `{file}`\n"));
    }

    markdown
}

pub fn render_wiki_markdown(structure: &WikiStructure) -> String {
    let mut markdown = String::new();
    markdown.push_str(&format!("# {}\n\n", structure.title));
    markdown.push_str(&format!("{}\n\n", structure.description));

    for section in &structure.sections {
        markdown.push_str(&format!("## {}\n\n", section.title));
        for page_id in &section.pages {
            if let Some(page) = structure.pages.iter().find(|page| &page.id == page_id) {
                markdown.push_str(&format!("- [{}](wiki/pages/{}.md)\n", page.title, page.id));
            }
        }
        markdown.push('\n');
    }

    markdown
}

File: python/src/deepwiki_mcp/wiki.py
"""Heuristic wiki scaffold and markdown rendering.

The heuristic plan is a fallback only — the MCP client (Claude) is expected
to call `save_wiki_plan` with a better-authored structure. We still need a
reasonable default so `plan_wiki` is callable without any LLM.
"""
from __future__ import annotations

from collections import OrderedDict

from .models import RepoScan, WikiPage, WikiSection, WikiStructure


def plan_wiki(scan: RepoScan, comprehensive: bool) -> WikiStructure:
    title = _repository_name(scan)
    title = f"{title} Wiki" if title else "Project Wiki"
    description = _build_description(scan)

    categorized = _categorize_files(scan.file_tree)

    sections: list[WikiSection] = []
    root_sections: list[str] = []
    pages: list[WikiPage] = []

    page_counter = 1
    section_counter = 1
    for category, file_paths in categorized.items():
        if not file_paths:
            continue
        section_id = f"section-{section_counter}"
        section_counter += 1
        page_id = f"page-{page_counter}"
        page_counter += 1

        title_case = _category_title(category)
        pages.append(
            WikiPage(
                id=page_id,
                title=title_case,
                description=f"Overview of the project's {title_case.lower()} implementation.",
                importance=_importance_for_category(category),
                file_paths=_trim_files(file_paths, comprehensive),
                related_pages=[],
                parent_section=section_id,
                content=None,
            )
        )
        sections.append(
            WikiSection(
                id=section_id,
                title=title_case,
                pages=[page_id],
                subsections=[],
            )
        )
        root_sections.append(section_id)

    if not pages:
        pages.append(
            WikiPage(
                id="page-1",
                title="Repository Overview",
                description="High-level documentation page for the repository.",
                importance="high",
                file_paths=_trim_files(list(scan.file_tree), comprehensive),
                related_pages=[],
                parent_section=None,
                content=None,
            )
        )

    all_ids = [page.id for page in pages]
    for page in pages:
        page.related_pages = [pid for pid in all_ids if pid != page.id][:3]

    return WikiStructure(
        id="wiki",
        title=title,
        description=description,
        sections=sections,
        root_sections=root_sections,
        pages=pages,
    )


def _categorize_files(files: list[str]) -> "OrderedDict[str, list[str]]":
    # Use sorted keys for deterministic output, matching Rust's BTreeMap.
    buckets: dict[str, list[str]] = {}
    for file in files:
        if file.startswith("tests/") or file.startswith("test/"):
            category = "testing"
        elif (
            file.startswith("src/app/")
            or file.startswith("src/components/")
            or file.endswith(".tsx")
            or file.endswith(".jsx")
        ):
            category = "frontend"
        elif file.startswith("api/") or file.endswith(".py"):
            category = "backend"
        elif (
            "config" in file
            or file.endswith(".json")
            or file.endswith(".toml")
            or file.endswith(".yaml")
            or file.endswith(".yml")
        ):
            category = "configuration"
        elif "docker" in file or "compose" in file or "deploy" in file:
            category = "deployment"
        elif file.endswith(".md"):
            category = "documentation"
        else:
            category = "architecture"
        buckets.setdefault(category, []).append(file)
    return OrderedDict(sorted(buckets.items()))


def _trim_files(files: list[str], comprehensive: bool) -> list[str]:
    limit = 12 if comprehensive else 6
    return files[:limit]


def _category_title(category: str) -> str:
    return {
        "frontend": "Frontend Components",
        "backend": "Backend Systems",
        "testing": "Testing Strategy",
        "configuration": "Configuration",
        "deployment": "Deployment and Infrastructure",
        "documentation": "Project Documentation",
    }.get(category, "System Architecture")


def _importance_for_category(category: str) -> str:
    if category in {"architecture", "backend", "frontend"}:
        return "high"
    if category in {"configuration", "deployment"}:
        return "medium"
    return "low"


def _repository_name(scan: RepoScan) -> str | None:
    name = scan.root_path.name
    return name or None


def _build_description(scan: RepoScan) -> str:
    if scan.readme_content:
        for raw in scan.readme_content.splitlines():
            line = raw.strip()
            if not line:
                continue
            if line.startswith("#") or line.startswith("---") or line.startswith("!["):
                continue
            if line.startswith("[!["):
                continue
            return line
    return "Repository documentation generated from source structure."


def render_page_markdown(page: WikiPage, scan: RepoScan) -> str:
    lines: list[str] = []
    lines.append("<details>")
    lines.append("<summary>Relevant source files</summary>")
    lines.append("")
    for file in page.file_paths:
        lines.append(f"- `{file}`")
    lines.append("</details>")
    lines.append("")
    lines.append(f"# {page.title}")
    lines.append("")
    lines.append(page.description)
    lines.append("")
    lines.append("## Summary")
    lines.append("")
    lines.append(
        f"This page was generated from the repository scan for `{scan.root_path}`. "
        "It currently serves as a structured stub and lists the source files that "
        "should drive future LLM-backed generation."
    )
    lines.append("")
    if scan.default_branch:
        lines.append("## Repository Metadata")
        lines.append("")
        lines.append(f"- Default branch: `{scan.default_branch}`")
        lines.append(f"- Scanned files: `{len(scan.file_tree)}`")
        lines.append("")
    lines.append("## Source Inventory")
    lines.append("")
    for file in page.file_paths:
        lines.append(f"- `{file}`")
    return "\n".join(lines) + "\n"


def render_wiki_index(structure: WikiStructure) -> str:
    lines: list[str] = []
    lines.append(f"# {structure.title}")
    lines.append("")
    lines.append(structure.description)
    lines.append("")
    page_by_id = {page.id: page for page in structure.pages}
    for section in structure.sections:
        lines.append(f"## {section.title}")
        lines.append("")
        for page_id in section.pages:
            page = page_by_id.get(page_id)
            if page is not None:
                lines.append(f"- [{page.title}](wiki/pages/{page.id}.md)")
        lines.append("")
    return "\n".join(lines) + "\n"
